### PR TITLE
Add support for GitSCM sources

### PR DESCRIPTION
**Pull request by uhafner**
Add support for `GitSCM` jobs. Obtains the required context from the latest build of FreeStyle jobs or pipelines. I refactored the existing context into a base class and a sub class to avoid code duplication. Additionally I removed the `run` field and initialized the corresponding fields in the constructor accordingly.

Example Results:
- https://github.com/uhafner/codingstyle/pull/198 (recorded from a local pipeline using GitSCM)
- https://github.com/uhafner/codingstyle/runs/988813405 (recorded from a local FreeStyle job using GitSCM)

Open points:
- [x] Requires that the `TaskListener` is never `null` (requires a upstream change in the API).
- [x] What should we do if there is no `run` available yet, then obtaining the SHA is not possible for those jobs. Currently publishing is just skipped.  
- [x] There is a Guava version conflict with Wiremock  and Jenkins core (required due to the new dependencies).
- [x] There are still some paths (maybe already in the existing code) that may lead to a NPE.
- [ ] There are still some paths that may lead to an exception during publishing (but this could be handled if we then would return a `NullPublisher`.

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/checks/github/GitSCMChecksContext.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import org.apache.commons.lang3.StringUtils;
+import org.jenkinsci.plugins.displayurlapi.DisplayURLProvider;
+
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.plugins.git.GitSCM;
+import hudson.plugins.git.UserRemoteConfig;
+
+/**
+ * Provides check properties that should be resolved  Jenkins job.
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/checks/github/GitSCMChecksContext.java:15]: this doesn't quite make sense to me is there a word missing?

**Comment by XiongKezhi** [comment]
thank you very much!!  It's a big enhancement, I'll take this weekend to review it.

**Comment by uhafner** [comment]
It's not ready for a review... It needs some more time, I'll convert from draft if it is working on my machine.
I basically created a parent class and copied the SCM code from my forensics plugin 😄 

What I am not sure on how to do: how do you get a HEAD hash in the current implementation if there is no build yet? Is it set by the branch source plugin automatically in the project even if there is no build yet? In freestyle builds a hash is only available after the first checkout.

**Comment by XiongKezhi** [comment]
> It's not ready for a review... It needs some more time, I'll convert from draft if it is working on my machine.
> I basically created a parent class and copied the SCM code from my forensics plugin 😄
> 
> What I am not sure on how to do: how do you get a HEAD hash in the current implementation if there is no build yet? Is it set by the branch source plugin automatically in the project even if there is no build yet? In freestyle builds a hash is only available after the first checkout.

the hash is fetched from github, see https://github.com/jenkinsci/github-checks-plugin/blob/4ecc8623b06d99d5f029b66927438554fdd6a467/src/main/java/io/jenkins/plugins/checks/github/GitHubSCMFacade.java#L74. However, one problem we found today is that since we are always fetching the head commit, so if the first one finishes later than the second one, it also uses the second one's commit to try to complete its own check, but it will end up updating the second one's check

**Comment by uhafner** [comment]
I see. I finally managed it to get it initialized for a Freestyle job. However it fails with:
```
[GitHub Checks] Using GitSCM repository 'jenkinsci/analysis-model' for GitHub checks
Failed Publishing GitHub checks: org.kohsuke.github.GHFileNotFoundException: https://api.github.com/repos/jenkinsci/analysis-model/check-runs {"message":"Resource not accessible by integration","documentation_url":"https://docs.github.com/rest/reference/checks#create-a-check-run"}
```

**Comment by XiongKezhi** [comment]
> I see. I finally managed it to get it initialized for a Freestyle job. However it fails with:
> 
> ```
> [GitHub Checks] Using GitSCM repository 'jenkinsci/analysis-model' for GitHub checks
> Failed Publishing GitHub checks: org.kohsuke.github.GHFileNotFoundException: https://api.github.com/repos/jenkinsci/analysis-model/check-runs {"message":"Resource not accessible by integration","documentation_url":"https://docs.github.com/rest/reference/checks#create-a-check-run"}
> ```

So you created your own GitHub app and install it on the org's repository?

**Comment by XiongKezhi** [comment]
> However, one problem we found today is that since we are always fetching the head commit, so if the first one finishes later than the second one, it also uses the second one's commit to try to complete its own check, but it will end up updating the second one's check.

And the solution would be simple in thought: just resolve the commit from build instead of fetch from remote after checkout stage.

So instead of listening to the job started, we now have to listen to checkout. cc @timja

**Comment by uhafner** [comment]
> So you created your own GitHub app and install it on the org's repository?

Of course, that was nonsense what I did. I created an app in my account but tried to publish in jenkinsci. I need to use my own repository to build. (Maybe it would make sense that we would catch such an error.)

**Comment by uhafner** [comment]
Ok, now I got it working for a freestyle build: https://github.com/uhafner/codingstyle/runs/988813405

🎉

**Comment by uhafner** [comment]
Hmm, in pipelines the environment variable GIT_COMMIT is not set, it seems that I need to resolve the SHA using the ˋGitClientˋ...

**Comment by XiongKezhi** [comment]
> Hmm, in pipelines the environment variable GIT_COMMIT is not set, it seems that I need to resolve the SHA using the ˋGitClientˋ...

One more thing, I'll change the listener in checks-api-plugin to listen to `checkout` instead of `onStarted` since checkout started later than `onStarted`, that would help the branch source implementation, hope that can help you as well.

File: src/main/java/io/jenkins/plugins/checks/github/GitHubSCMFacade.java [hunk 2]
```diff
@@ -31,6 +49,48 @@
...
+    }
+    
+    /**
+     * Finds the {@link GitSCM} used by the {@code run}.
+     *
+     * @param job
+     *         the run to get the SCM from 
+     * @return the found GitSCM or empty
+     */
+    public Optional<GitSCM> findGitSCM(final Job<?, ?> job) {
```
[the remark this answers is not in the document]
Comment (uhafner) [reply src/main/java/io/jenkins/plugins/checks/github/GitHubSCMFacade.java]: This one is used

**Comment by uhafner** [comment]
The tests fail on my machine with 
```
[ERROR] io.jenkins.plugins.checks.github.GitHubChecksPublisherITest.shouldLogChecksParametersIfExceptionHappensWhenPublishChecks  Time elapsed: 1.736 s  <<< ERROR!
java.lang.NoClassDefFoundError: com/google/common/collect/FluentIterable
```

File: src/main/java/io/jenkins/plugins/checks/github/SCMFacade.java [hunk 3]
```diff
@@ -19,6 +19,7 @@
 import jenkins.scm.api.SCMRevision;
 import jenkins.scm.api.SCMSource;
 import jenkins.triggers.SCMTriggerItem;
+import jline.internal.Nullable;
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/checks/github/SCMFacade.java:22]: this looks like a strange annotation, normally you would use `CheckForNull` from spotbugs

Comment (uhafner) [reply src/main/java/io/jenkins/plugins/checks/github/SCMFacade.java:22]: Yes, that's why Spotbugs still does find a problem...

**Comment by XiongKezhi** [comment]
> > > There is a Guava version conflict with Wiremock and Jenkins core (required due to the new dependencies).
> 
> > Fixed? Passed my local build
> 
> The tests fail on my machine with
> 
> ```
> [ERROR] io.jenkins.plugins.checks.github.GitHubChecksPublisherITest.shouldLogChecksParametersIfExceptionHappensWhenPublishChecks  Time elapsed: 1.736 s  <<< ERROR!
> java.lang.NoClassDefFoundError: com/google/common/collect/FluentIterable
> ```

the CI passed as well

**Comment by XiongKezhi** [comment]
> > > > There is a Guava version conflict with Wiremock and Jenkins core (required due to the new dependencies).
> > 
> > 
> > > Fixed? Passed my local build
> > 
> > 
> > The tests fail on my machine with
> > ```
> > [ERROR] io.jenkins.plugins.checks.github.GitHubChecksPublisherITest.shouldLogChecksParametersIfExceptionHappensWhenPublishChecks  Time elapsed: 1.736 s  <<< ERROR!
> > java.lang.NoClassDefFoundError: com/google/common/collect/FluentIterable
> > ```
> 
> the CI passed as well

yeah, now I reproduced that after remove `@Ignore`, didn't notice that before.



> > > > There is a Guava version conflict with Wiremock and Jenkins core (required due to the new dependencies).
> > 
> > 
> > > Fixed? Passed my local build
> > 
> > 
> > The tests fail on my machine with
> > ```
> > [ERROR] io.jenkins.plugins.checks.github.GitHubChecksPublisherITest.shouldLogChecksParametersIfExceptionHappensWhenPublishChecks  Time elapsed: 1.736 s  <<< ERROR!
> > java.lang.NoClassDefFoundError: com/google/common/collect/FluentIterable
> > ```
> 
> the CI passed as well

Reproduced that after remove `@Ignore`

**Comment by uhafner** [comment]
Jenkins uses a very old version of Guava, this is the reason that I am not using Guava in my sources in order to avoid conflicts.   Is there a version of Wiremock available that uses shaded libraries?

**Comment by XiongKezhi** [comment]
> Jenkins uses a very old version of Guava, this is the reason that I am not using Guava in my sources in order to avoid conflicts. Is there a version of Wiremock available that uses shaded libraries?

Lucky that wiremock also depends on a very old guava version (20.0, see https://mvnrepository.com/artifact/com.github.tomakehurst/wiremock/2.27.1), I added it in the pom and it works:

    <dependency>
      <groupId>com.google.guava</groupId>
      <artifactId>guava</artifactId>
      <version>${guava.version}</version>
      <scope>test</scope>
    </dependency>

**Review by XiongKezhi — APPROVED** [review]

**Pull request merged**